### PR TITLE
Extend `DecomposeClassicalExp` to handle `ClExprOp`

### LINKED ISSUE #1581: `DecomposeClassicalExp()` fails for constant conditions
```python
from pytket import Circuit
from pytket.circuit.logic_exp import BitZero, BitOne
from pytket.passes import DecomposeClassicalExp

circuit = Circuit(1)
circuit.X(0, condition=BitZero())
DecomposeClassicalExp().apply(circuit)
```

Output:

```
Traceback (most recent call last):
  File "/home/alec/r/tket/pytket/pytket/circuit/decompose_classical.py", line 334, in _decompose_expressions
    comp_bit = bit_recursive_walk(pred_exp, kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alec/r/tket/pytket/pytket/circuit/decompose_classical.py", line 230, in recursive_walk
    arg1, arg2 = exp.args
    ^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 0)
```

### PR DESCRIPTION
**Pull request by cqc-alec**
Also fix #1581 by adding support for constant expressions (in `ClExprOp`).

### CODE REVIEW

File: pytket/tests/qasm_test_files/test18_output.qasm
Comment (cqc-alec) [inline pytket/tests/qasm_test_files/test18_output.qasm:1]: Note that this reverts to the file as it was before #1628 .

File: pytket/tests/qasm_test.py [hunk 1]
```diff
@@ -464,14 +465,18 @@ def test_extended_qasm() -> None:
 
     assert circuit_to_qasm_str(c2, "hqslib1")
 
-    assert not DecomposeClassicalExp().apply(c)
+    with pytest.raises(DecomposeClassicalError):
```
Comment (cqc-alec) [inline pytket/tests/qasm_test.py:468]: This reverts to the test as it was before #1628 .

File: pytket/pytket/circuit/decompose_classical.py [hunk 2]
```diff
@@ -343,6 +478,31 @@ def _decompose_expressions(circ: Circuit) -> tuple[Circuit, bool]:
...
+            expr: ClExpr = wexpr.expr
+            bit_posn = wexpr.bit_posn
+            reg_posn = wexpr.reg_posn
+            output_posn = wexpr.output_posn
+            assert len(output_posn) > 0
+            output0 = args[output_posn[0]]
+            assert isinstance(output0, Bit)
+            out_var: Variable = (
+                BitRegister(output0.reg_name, len(output_posn))
+                if has_reg_output(expr.op)
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply pytket/pytket/circuit/decompose_classical.py:494]: (for example when using temp bits?)

[the remark this answers is not in the document]
Comment (cqc-alec) [reply pytket/pytket/circuit/decompose_classical.py:494]: The check at line 380 would catch this case and throw an exception. We do not allow this pass to be run on circuits with non-register-aligned `ClExprOp` (even though they are valid pytket circuits). I think removing this restriction would be possible, but would make decomposition more complicated, so prefer to keep it unless there is a real need.

Temporary (scratch) bits would not normally form part of a `ClExprOp`. They are introduced when decomposing it into elementary classical operations.

**Review by cqc-melf — APPROVED** [review]

**Pull request merged**